fiona/ogrext.pyx

```diff
@@ -525,16 +521,53 @@ cdef class Session:
         self.cogr_ds = NULL
 
     def get_fileencoding(self):
+        """DEPRECATED"""
         return self._fileencoding
 
-    def get_internalencoding(self):
-        if not self._encoding:
-            fileencoding = self.get_fileencoding()
-            self._encoding = (
-                OGR_L_TestCapability(
-                    self.cogr_layer, OLC_STRINGSASUTF8) and
-                'utf-8') or fileencoding
-        return self._encoding
+    def _get_fallback_encoding(self):
+        """Determine a format-specific fallback encoding to use when using OGR_F functions
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        str
+
+        """
+        if "Shapefile" in self.get_driver():
+            return 'iso-8859-1'
+        else:
+            return locale.getpreferredencoding()
+
+
+    def _get_internal_encoding(self):
+        """Determine the encoding to use when use OGR_F functions
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        str
+
+        Notes
+        -----
+        If the layer implements RFC 23 support for UTF-8, the return
+        value will be 'utf-8' and callers can be certain that this is
+        correct.  If the layer does not have the OLC_STRINGSASUTF8
+        capability marker, it is not possible to know exactly what the
+        internal encoding is and this method returns best guesses. That
+        means ISO-8859-1 for shapefiles and the locale's preferred
+        encoding for other formats such as CSV files.
+
+        """
+        if OGR_L_TestCapability(self.cogr_layer, OLC_STRINGSASUTF8):
+            return 'utf-8'
+        else:
+            return self._fileencoding or self._get_fallback_encoding()
```

Comment:
This function returns the name of the codec we now use when going between Python and OGR field names and values.